interfaces/builtin/usb_gadget.go

```diff
@@ -29,25 +48,199 @@ const usbGadgetBaseDeclarationSlots = `
     deny-auto-connection: true
 `
 
-const usbGadgetConnectedPlugAppArmor = `
-# https://www.kernel.org/doc/Documentation/usb/gadget_configfs.txt
-# Allow creating new gadgets under usb_gadget, which is creating
-# new directories
-/sys/kernel/config/usb_gadget/ rw,
-# Allow creating sub-directories, symlinks and files under those
-# directories
-/sys/kernel/config/usb_gadget/** rw,
-
-# Allow access to UDC
-/sys/class/udc/ r,
+const usbGadgetConnectedPlugSecComp = `
+# Description: Allow mount and umount syscall access. No filtering here, as we
+# rely on AppArmor to filter the mount operations.
+mount
+umount
+umount2
 `
 
+// usbGadgetInterface allows creating transient and persistent mounts
+type usbGadgetInterface struct {
+	commonInterface
+}
+
+type ffsMountInfo struct {
+	name       string
+	where      string
+	persistent bool
+}
+
+func (mi *ffsMountInfo) validate() error {
+	// for ffs the name is the name of the function, which is not a path so we
+	// just need to ensure it doesn't contain any AppArmor regex characters.
+	if err := validateNoAppArmorRegexpWithError(`cannot use usb-gadget "name" attribute`, mi.name); err != nil {
+		return err
+	}
+
+	// Reuse the where validation from mount-control, as the semantics are the same
+	if err := validateWhereAttr(mi.where); err != nil {
+		return err
+	}
+
+	// Until we have a clear picture of how this should work, disallow creating
+	// persistent mounts into $SNAP_DATA
+	if mi.persistent && strings.HasPrefix(mi.where, "$SNAP_DATA") {
+		return errors.New(`mount-control "persistent" attribute cannot be used to mount onto $SNAP_DATA`)
+	}
+
+	return nil
+}
+
+func ffsMounts(plug interfaces.Attrer) ([]map[string]any, error) {
+	var mounts []map[string]any
+	err := plug.Attr("ffs-mounts", &mounts)
+	if err != nil && !errors.Is(err, snap.AttributeNotFoundError{}) {
+		return nil, mountAttrTypeError
+	}
+	return mounts, nil
+}
+
+func enumerateFFSMounts(mounts []map[string]any, fn func(mi *ffsMountInfo) error) error {
```

Comment:
Could this be a new-style iter.Seq like function that we use without the syntax sugar? The difference is that you get `bool` instead of `error` at the end but otherwise it behaves very much the same. When we update to 2.24 we get the super clean usage as well (it doesn't affect the definition, hence my comment).

Have a look at https://github.com/canonical/snapd/pull/16007/files#diff-44ca0b0f4c192b98966199af6e778ae88ab45bf18a8621125f9fc1874e20c7d7R102-R104 for an example from VFS.

This is just a suggestion, not a blocker.

Reply:
I think this is pretty cool - I updated the code to implement `iter.Seq2`.